Refactor of commands and flag parsing

PhilipGough opened the pull request:
<!--
    Keep PR title verbose enough and add prefix telling
    about what components it touches e.g "query:" or ".*:"
-->
Introduce config structs for sidecar component.
<!--
    Don't forget about CHANGELOG!

    Changelog entry format:
    - [#<PR-id>](<PR-URL>) Thanos <Component> ...

    <PR-id> Id of your pull request.
    <PR-URL> URL of your PR such as https://github.com/thanos-io/thanos/pull/<PR-id>
    <Component> Component affected by your changes such as Query, Store, Receive.
-->

* [x] Change is not relevant to the end user.

## Changes

As per #2248  - this change parses flags into structs before passing into the sidecar run func

kakkoyun commented on cmd/thanos/sidecar.go:
```diff
@@ -44,100 +43,53 @@ import (
 }
 
 func runSidecar(
+	conf sidecarComponentConfig,
```
I  believe it'd be better to put conf as the last argument when initializing a component. If you check the code base, you'd see injected dependencies come first.

kakkoyun commented on cmd/thanos/config.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+type grpcConfig struct {
+	bindAddress    *string
```
I think when we should have the values in the config rather than pointers. It'd be cleaner in that way, otherwise, we'd end up lots of value extactions.
We can leave pointers if they're optional.

bwplotka commented:
Started: https://github.com/bwplotka/flagarize

Hopefully can finish something this weekend :muscle:  :heart: 

![](https://media0.giphy.com/media/1hMdWxlfKh13NOMRvP/giphy.gif)

PhilipGough commented:
@kakkoyun, @bwplotka - thanks for the feedback. I'll be away for a few days so likely the middle of next week before I get back to take a look at this properly.

@bwplotka linked library looks promising. I can look at integrating that if you think it is cleaner to define all the details as struct tags?

bwplotka commented:
Which exactly link you mean?

On Sat, 14 Mar 2020, 15:16 Philip Gough, <notifications@github.com> wrote:

> @kakkoyun <https://github.com/kakkoyun>, @bwplotka
> <https://github.com/bwplotka> - thanks for the feedback. I'll be away for
> a few days so likely the middle of next week before I get back to take a
> look at this properly.
>
> @bwplotka <https://github.com/bwplotka> linked library looks promising. I
> can look at integrating that if you think it is cleaner to define all the
> details as struct tags?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/thanos-io/thanos/pull/2267#issuecomment-599077849>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVA3OZCN5KH5D7KJZEUUF3RHONWLANCNFSM4LHD4QLQ>
> .
>

PhilipGough commented:
@bwplotka sorry, this lib https://github.com/octago/sflags

bwplotka commented:
Yes, I researched it a bit and I don't like it really. It is too generic
and will not work with our custom flag structs. That's why I started
https://github.com/bwplotka/flagarize. Should be ready when you are back!
(:

On Sat, 14 Mar 2020, 16:34 Philip Gough, <notifications@github.com> wrote:

> @bwplotka <https://github.com/bwplotka> sorry, this lib
> https://github.com/octago/sflags
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/thanos-io/thanos/pull/2267#issuecomment-599093054>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVA3O7KG5BZNAOFHPZ6VM3RHOWZ5ANCNFSM4LHD4QLQ>
> .
>

bwplotka commented:
I released v0.9.0 of https://github.com/bwplotka/flagarize... should we try to adopt it? 

I already refactored Prometheus for it: https://github.com/prometheus/prometheus/pull/7026

GiedriusS commented:
> I released v0.9.0 of https://github.com/bwplotka/flagarize... should we try to adopt it?
> 
> I already refactored Prometheus for it: [prometheus/prometheus#7026](https://github.com/prometheus/prometheus/pull/7026)

Thanks for the blog post - the idea itself looks good to me. Is this ready for review? Seems like the CI doesn't like this.

bwplotka commented:
I believe this is not. Philip wanted to move to flagarize.

On Mon, 23 Mar 2020 at 15:11, Giedrius Statkevičius <
notifications@github.com> wrote:

> I released v0.9.0 of https://github.com/bwplotka/flagarize... should we
> try to adopt it?
>
> I already refactored Prometheus for it: prometheus/prometheus#7026
> <https://github.com/prometheus/prometheus/pull/7026>
>
> Thanks for the blog post - the idea itself looks good to me. Is this ready
> for review? Seems like the CI doesn't like this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/thanos-io/thanos/pull/2267#issuecomment-602663123>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVA3OZDK5HVXFCRMBQ2SGLRI53ZLANCNFSM4LHD4QLQ>
> .
>

PhilipGough commented:
@GiedriusS , @bwplotka hey, no not ready for review. I just updated one of the components for feedback on how it would look. I'll introduce flagarize and updated the other components asap

bwplotka commented:
Any update? (:

bwplotka commented on cmd/thanos/config.go:
```diff
@@ -0,0 +1,61 @@
+	connection connectionConfig
+	reloader   reloaderConfig
+	shipper    shipperConfig
+	objStore   *extflag.PathOrContent
```
This needs some work or we should use `flagarize` PathOrContent instead (and remove our flag)

bwplotka commented on cmd/thanos/sidecar.go:
```diff
@@ -45,70 +46,31 @@ import (
 	minTime := thanosmodel.TimeOrDuration(cmd.Flag("min-time", "Start of time range limit to serve. Thanos sidecar will serve only metrics, which happened later than this value. Option can be a constant time in RFC3339 format or time duration relative to current time, such as -1d or 2h45m. Valid duration units are ms, s, m, h, d, w, y.").
 		Default("0000-01-01T00:00:00Z"))
 
 	m[component.Sidecar.String()] = func(g *run.Group, logger log.Logger, reg *prometheus.Registry, tracer opentracing.Tracer, _ <-chan struct{}, _ bool) error {
```
I think we can now totally remove this item. Instead of registration we can invoke run directly I think :+1:

brancz commented on cmd/thanos/config.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+type grpcConfig struct {
+	BindAddress    string         `flagarize:"name=Grpc-address|help=Listen ip:port address for gRPC endpoints (StoreAPI). Must be routable from other components.|default=0.0.0.0:10901"`
```
I'm not a huge fan of the struct tags in this way. We're throwing away all the type safety and language characteristics that we could previously use with kingpin. Why can't we capture the information kept in the struct tags here in real structs?

PhilipGough replied:
@brancz Thats the approach I took in the initial commit https://github.com/thanos-io/thanos/pull/2267/commits/2562b321fbd8d7dd6073f65299272c67d824c1af

I just integrated the library after the first review as suggested but am happy to revert if thats the consensus

brancz replied:
I think I would personally prefer that approach.

PhilipGough replied:
@brancz - I can let yourself and @bwplotka battle that one out in that case :D

bwplotka replied:
I can see lack of type safety, but the advantage is huge - we have finally a way to tie struct filed directly, in the same readable place to a flag. In the same way we can do at some point with `yaml` or json parsing if we choose to have config based configuration. 

Also we literally don't need type safety here, because we always check on our CI the issue on the very start by doing `thanos --help` on every command. (make docs). So overall it does not matter if the error on registering flags is in compile-time or runtime. That's why, while I love type safety, here I am able to sacrifice is for a very good reason.

bwplotka replied:
If you still disagree it would be nice to have opinions from other maintainers as well e.g: @squat @kakkoyun @povilasv @GiedriusS

bwplotka replied:
Overall yea, let's discuss and ensure we all agree here. I can see this being "bit surprising" approach (:

povilasv replied:
I'm also not a big fan of struct tags, but I think we can work around it by having good linter in place, which would complain if it can't parse the tag and explain whats wrong.

pracucci replied:
_My very personal perspective:_

To my understanding, the plan is to introduce a linter to do the compiler job just because we want to specify CLI flags as struct tags instead of code. Why do we want it? Because "we have finally a way to tie struct filed directly, in the same readable place to a flag".

As you probably know, in Cortex we basically achieve the same result attaching a `RegisterFlags()` function to each config struct and we recursively call `RegisterFlags()` on sub-config structs. Despite there's more code involved, I personally find it a safer approach (and avoids to have to spend further time writing a linter).

Example:
https://github.com/cortexproject/cortex/blob/9245c3120d7b16117aaffcb383a6347d002475d1/pkg/compactor/compactor.go#L52-L61

bwplotka replied:
Awesome, thanks for the feedback. Looks like flagarize is not the best option here. I can totally agree with your points. As always there is some power in type safety and more important: autocompletion. (: 

@PhilipGough sorry for taking this back and forth and thanks for giving us the opportunity to see how flagarize would work in Thanos (:

To me looks like we settled on just:

```go
type ...Config struct {
    httpAddress string
}
c := ...Config{}
cmd.Flag( "name", "help").Default().StringVar(&c.httpAddress)
``` 

It's almost similar to what Cortex has except one thing: I am pretty sure we should never define flags outside main package.

BTW,  I only wished @brancz would have the same opinion for other languages and patterns... ;p

povilasv replied:
I've sort of like flagarize, as we would have only one place where we define the config, instead of flags & struct.
The linter thing doesn't have to happen now, so the PR & approach to me looks good.

brancz replied:
I actually quite like the pattern in Cortex.

bwplotka replied:
Let's discuss it more I feel like we have 3 options already (: 

1. Flagarize
2. Struct with typed flag registration in main
3.  Struct with typed flag registration in each component.

I believe flag should never be part of package. We can then limit the surfaces we should care  for each package. If suddenly add flags we add much more things we need to care for in terms of complexity, compatibility etc on top of what we already have with metrics and logs... Imagine e.g reloader package is used in Thanos and X projects (e.g part of prometheus-operator is using it). If we would add flag registration to reloader flag we introduce huge pain in renaming, what should be default should be hidden or not? Quite a pain ): 

So for me no 2 wins, we could think about flagarize more, but looks like 2 is good step in this direction we all are comfortable to take IMO (:

brancz replied:
I'm happy with that as a first step.

pracucci replied:
> I believe flag should never be part of package

That's what happen with flagarize too. Being coupled with the config struct, whenever you move it to a package flags definition moves along it.

> 2. Struct with typed flag registration in main

Even if this option misses the point of "we have finally a way to tie struct filed directly, in the same readable place to a flag", I still see it valuable to move config to `struct` 👍

bwplotka replied:
> That's what happen with flagarize too

Well that's the point we should never use even flagarize outside of main (:

povilasv replied:
I have a bit different opinion, one thing having multiple structs / config options in packages gives us that you can add build tags which could be compiled for embedded stuff. 

For example I'm using s3 and I don't need all that azure stuff and it's dependencies. This would make binary size way smaller, etc. 

One example of this pattern is node exporter actually does command line flags in the different packages and it's fine:

https://github.com/prometheus/node_exporter/blob/master/collector/systemd_linux.go#L41-L48

https://github.com/prometheus/node_exporter/blob/master/collector/filesystem_linux.go#L38-L41

bwplotka replied:
Interesting. @povilasv you are right but we can already do that as we hide all the plugins with the config pathOrContent yaml (:

bwplotka commented on cmd/thanos/config.go:
```diff
@@ -0,0 +1,63 @@
+	ignoreBlockSize bool
+}
+
+type sidecarComponentConfig struct {
```
Ideally all of those would be in each command file (:

bwplotka commented on cmd/thanos/flags.go:
```diff
@@ -73,3 +75,99 @@ func regSelectorRelabelFlags(cmd *kingpin.CmdClause) *extflag.PathOrContent {
+		uploadCompacted: *cmd.Flag("shipper.upload-compacted",
+			"If true sidecar will try to upload compacted blocks as well. Useful for migration purposes. Works only if compaction is disabled on Prometheus. Do it once and then disable the flag when done.").
+			Default("false").Bool(),
+		ignoreBlockSize: *cmd.Flag("shipper.ignore-unequal-block-size",
```
So to make it work we need opposite: We need `BoolVar(&ignoreBlockSize)`.

How it works is that you register variables here first. Kingpin either returns address of the variable that the value will be AFTER parsing or it allows us to tell kingpin where is the address (using `*Var()` methods). Let's do the later so our config struct will fill (:

bwplotka commented on cmd/thanos/flags.go:
```diff
@@ -73,3 +75,99 @@ func regSelectorRelabelFlags(cmd *kingpin.CmdClause) *extflag.PathOrContent {
+	}
+}
+
+func parseFlagsToShipperConf(cmd *kingpin.CmdClause) shipperConfig {
```
We can follow Cortex way of doing thing and have this as just `RegisterFlag` method on shipperConfig struct directly.


```suggestion
func (c *shipperConfig) RegisterFlag(cmd *kingpin.CmdClause) {
```

bwplotka commented on cmd/thanos/sidecar.go:
```diff
@@ -457,3 +398,39 @@ func (s *promMetadata) Timestamps() (mint int64, maxt int64) {
+}
+
+func newSidecarConfig(cmd *kingpin.CmdClause) *sidecarConfig {
+	var grpcConf grpcConfig
```
We don't need any of this var definition, we can use directly sidecar config (: 

I think what would look better is just RegisterFlag method for sidecar.


Then we can create sidecar struct and do `RegisterFlags`. Then in that method we can run all RegisterFlags of each field for this struct.  Those method has to have pointer receivers to make a difference.

bwplotka replied:
It's bit repetitive, but good enough (:

PhilipGough commented on cmd/thanos/sidecar.go:
```diff
@@ -458,3 +400,29 @@ func (s *promMetadata) Timestamps() (mint int64, maxt int64) {
+	sc.reloader.registerFlag(cmd)
+	sc.objStore = *regCommonObjStoreFlags(cmd, "", false)
+	sc.shipper.registerFlag(cmd)
+	sc.limitMinTime = *thanosmodel.TimeOrDuration(cmd.Flag("min-time", "Start of time range limit to serve. Thanos sidecar will serve only metrics, which happened later than this value. Option can be a constant time in RFC3339 format or time duration relative to current time, such as -1d or 2h45m. Valid duration units are ms, s, m, h, d, w, y.").
```
@bwplotka are you happy with how this and the one above it are being parsed at this stage?

PhilipGough replied:
actually I see `limitMinTime` isn't being set correctly yet

bwplotka commented on cmd/thanos/config.go:
```diff
@@ -0,0 +1,127 @@
+	return pc
+}
+
+type connectionConfig struct {
```
```suggestion
type connConfig struct {
```

bwplotka commented on cmd/thanos/sidecar.go:
```diff
@@ -458,3 +400,29 @@ func (s *promMetadata) Timestamps() (mint int64, maxt int64) {
+	limitMinTime thanosmodel.TimeOrDurationValue
+}
+
+func (sc *sidecarConfig) registerFlag(cmd *kingpin.CmdClause) *sidecarConfig {
```
Looking forward for some tool for this, but let's do in separate PR (:

bwplotka commented:
So much cleaner :heart: 

@brancz @kakkoyun @squat @GiedriusS looks good? (:

PhilipGough commented on cmd/thanos/sidecar.go:
```diff
@@ -44,112 +44,52 @@ import (
 }
 
 func runSidecar(
+	conf sidecarConfig,
```
Yep, I had actually fixed this previously but missed it when I reverted back. will fix

bwplotka commented:
:muscle: @PhilipGough Thanks!
